### DISCUSSION TITLE
capturing information in the lamba of the Insert function

### DISCUSSION THREAD

**Discussion by grbba** [daniele77/cli#105] [Q&A]
Hello,
i must have a fundamental misunderstanding of C++ lambdas i guess but nevermind. I have a vector of structs which hold 
the name, vector of the parameter descriptions and a help string over which I iterate to insert them into the menu.
In the lambda there is one call to execute the command which shall take the name of the menu item and in the execution function display them. ( As i will have quite some variable parameter number commands and not wanting to have n times the same command in the list for each possible parameter number as well as having too much copy paste programming to do for the Insert Commands ). I'd like to capture the name of the command in the executor so i know what to do next.

```
...
void executeMenu(std::ostream &out, std::string mi, std::vector<std::string> params)
{
  out << "execute menu item: " << mi << '\n';
}
...
  auto rootMenu = make_unique<Menu>("cli");

  for (auto c : cmd)
  {
      rootMenu->Insert(
      c.name,
      c.desc,
      [&](std::ostream &out, const std::vector<std::string> params ) {
        executeMenu(out, c.name, params);
      },
      c.help);
  }
 ...
```
That works fine for defining the commands in the menu. They show up as expected but the name printed in the executeMenu() above is always the name of the last command entered through the loop ... and when executed several times nothing shows up again as you can see below

```
Welcome to the DCC++ EX Commandline Interface!

cli> help
Commands available:
 - help
        This help message
 - exit
        Quit the session
 - do1 <[ack|wifi|ethernet|cmd|wit]> <[on|off]>
        help for do1
 - do2 <[parameter 1]> <[on|off]>
        help for do2
cli> do1
execute menu item: do2
cli> do2
execute menu item: 
cli> do1
execute menu item: 
cli> exit
Closing App...
Goodbye and thanks for all the fish.
grbba@aplfa src % 
```

I guess i have just something very simple to do to get this going but i can't get my head around this with my little programming experience

The struct which contains everything i need is build from JSON strings so that i can add easily new menu items by just editing their description and providing the function to execute what they should do. All the wiring is taken care of ...

Doing it this way where m1 m1desc and m1help are individually instantiated global variables works fine:

```
std::string m1 = "do1";
std::vector<std::string> m1desc = {"[ack|wifi|ethernet|cmd|wit]", "[on|off]"};
std::string m1help = "help for do1";

  rootMenu->Insert(
      m1,
      m1desc,
      [&](std::ostream &out, const std::vector<std::string> params ) {
        executeMenu(out, m1, params);
      },
      m1help);
```

Full file from which i tested: 
[main.cpp.txt](https://github.com/daniele77/cli/files/6012969/main.cpp.txt)

Thanks a lot for your help 
//

**Comment by grbba** [comment]
Ok first sorry for the stupid question above.. I have found how to do it. The trick is basically capturing not the references but the values i.e. replacing the & by = in the lamba expression in the for loop. That way it works. Is it the right way to do this sort of things though ? No idea

//grbba

**Reply by daniele77** [reply]
Ok, you found the problem: the capturing by reference.

As for your final question: if you want to provide a JSON description of your commands I guess this is a good way to do it. Obviously, to have a JSON description, your command actions must be pretty uniform (e.g., each command results in sending a message over a TTY).
